app/code/Magento/SalesRule/Model/ResourceModel/Report/Rule.php

```diff
@@ -5,43 +5,55 @@
+use Magento\Framework\Stdlib\DateTime\Timezone\Validator;
+use Magento\Framework\Stdlib\DateTime\TimezoneInterface;
+use Magento\Reports\Model\Flag;
+use Magento\Reports\Model\FlagFactory;
+use Magento\Reports\Model\ResourceModel\Report\AbstractReport;
+use Magento\SalesRule\Model\ResourceModel\Report\Rule\CreatedatFactory;
+use Magento\SalesRule\Model\ResourceModel\Report\Rule\UpdatedatFactory;
+use Psr\Log\LoggerInterface;
+use Zend_Db_Expr;
+
 /**
  * Rule report resource model
  *
  * @author      Magento Core Team <core@magentocommerce.com>
  */
-class Rule extends \Magento\Reports\Model\ResourceModel\Report\AbstractReport
+class Rule extends AbstractReport
 {
     /**
-     * @var \Magento\SalesRule\Model\ResourceModel\Report\Rule\CreatedatFactory
+     * @var CreatedatFactory
      */
     protected $_createdatFactory;
 
     /**
-     * @var \Magento\SalesRule\Model\ResourceModel\Report\Rule\UpdatedatFactory
+     * @var UpdatedatFactory
      */
     protected $_updatedatFactory;
 
     /**
-     * @param \Magento\Framework\Model\ResourceModel\Db\Context $context
-     * @param \Psr\Log\LoggerInterface $logger
-     * @param \Magento\Framework\Stdlib\DateTime\TimezoneInterface $localeDate
-     * @param \Magento\Reports\Model\FlagFactory $reportsFlagFactory
-     * @param \Magento\Framework\Stdlib\DateTime\Timezone\Validator $timezoneValidator
-     * @param \Magento\Framework\Stdlib\DateTime\DateTime $dateTime
-     * @param \Magento\SalesRule\Model\ResourceModel\Report\Rule\CreatedatFactory $createdatFactory
-     * @param \Magento\SalesRule\Model\ResourceModel\Report\Rule\UpdatedatFactory $updatedatFactory
+     * @param DbContext $context
+     * @param LoggerInterface $logger
+     * @param TimezoneInterface $localeDate
+     * @param FlagFactory $reportsFlagFactory
+     * @param Validator $timezoneValidator
+     * @param DateTime $dateTime
+     * @param CreatedatFactory $createdatFactory
+     * @param UpdatedatFactory $updatedatFactory
      * @param string $connectionName
      */
     public function __construct(
-        \Magento\Framework\Model\ResourceModel\Db\Context $context,
-        \Psr\Log\LoggerInterface $logger,
-        \Magento\Framework\Stdlib\DateTime\TimezoneInterface $localeDate,
-        \Magento\Reports\Model\FlagFactory $reportsFlagFactory,
-        \Magento\Framework\Stdlib\DateTime\Timezone\Validator $timezoneValidator,
-        \Magento\Framework\Stdlib\DateTime\DateTime $dateTime,
-        \Magento\SalesRule\Model\ResourceModel\Report\Rule\CreatedatFactory $createdatFactory,
-        \Magento\SalesRule\Model\ResourceModel\Report\Rule\UpdatedatFactory $updatedatFactory,
+        DbContext $context,
+        LoggerInterface $logger,
+        TimezoneInterface $localeDate,
+        FlagFactory $reportsFlagFactory,
+        Validator $timezoneValidator,
+        DateTime $dateTime,
+        CreatedatFactory $createdatFactory,
+        UpdatedatFactory $updatedatFactory,
```

Comment:
Why not promote these properties?
```suggestion
        protected readonly CreatedatFactory $_createdatFactory,
        protected readonly UpdatedatFactory $_updatedatFactory,
```